spec/consumer_resource_spec.rb

```diff
@@ -1553,6 +1551,82 @@ def verify_installed_pids(consumer, product_ids)
+    env2 = @cp.create_environment(@owner1['key'], random_string("env_id-2"), random_string("env_name-2"))
+
+    consumer = user_cp.register(random_string('system'), :system, nil, {}, nil, nil, [], [])
+    consumer_client = Candlepin.new(nil, nil, consumer['idCert']['cert'], consumer['idCert']['key'])
+
+    consumer_client.update_consumer({:environments => [{:name => env1.name}, {:name => env2.name}]})
+    consumer = @cp.get_consumer(consumer['uuid'])
+    expect(consumer.environments[0].name).to eq(env1.name)
+    expect(consumer.environments[1].name).to eq(env2.name)
+    # re-prioritized the environments
+    consumer_client.update_consumer({:environments => [{:name => env2.name}, {:name => env1.name}]})
+    consumer = @cp.get_consumer(consumer['uuid'])
+    expect(consumer.environments[0].name).to eq(env2.name)
+    expect(consumer.environments[1].name).to eq(env1.name)
+
+  end
+
+  it 'should update consumer with updated environments' do
+    user_cp = user_client(@owner1, random_string('billy'))
+    env1 = @cp.create_environment(@owner1['key'], random_string("env_id-1"), random_string("env_name-1"))
+    env2 = @cp.create_environment(@owner1['key'], random_string("env_id-2"), random_string("env_name-2"))
+    env3 = @cp.create_environment(@owner1['key'], random_string("env_id-3"), random_string("env_name-3"))
+
+
+    consumer = user_cp.register(random_string('system'), :system, nil, {}, nil, nil, [], [],
+      [{'id' => env1.id}, {'id' => env2.id}, {'id' => env3.id}])
+    consumer_client = Candlepin.new(nil, nil, consumer['idCert']['cert'], consumer['idCert']['key'])
+
+    expect(consumer.environments[0].name).to eq(env1.name)
+    expect(consumer.environments[1].name).to eq(env2.name)
+    expect(consumer.environments[2].name).to eq(env3.name)
+
+    # Removed the environment 2 (env_id-2).
+    consumer_client.update_consumer({:environments => [{:name => env1.name}, {:name => env3.name}]})
+    consumer = @cp.get_consumer(consumer['uuid'])
+
+    #It should maintain the lowering priority for the rest of other environments.
+    expect(consumer.environments[0].name).to eq(env1.name)
+    expect(consumer.environments[1].name).to eq(env3.name)
+  end
+
+  it 'should return exception when environment is specified more than once' do
+    user_cp = user_client(@owner1, random_string('billy'))
+    env1 = @cp.create_environment(@owner1['key'], random_string("env_id-1"), random_string("env_name-1"))
+    env2 = @cp.create_environment(@owner1['key'], random_string("env_id-2"), random_string("env_name-2"))
+
+    consumer = user_cp.register(random_string('system'), :system, nil, {}, nil, nil, [], [])
+    consumer_client = Candlepin.new(nil, nil, consumer['idCert']['cert'], consumer['idCert']['key'])
+
+    expect {
+      consumer_client.update_consumer({:environments => [{:name => env1.name}, {:name => env2.name}, {:name => env1.name}]})
```

Comment:
Something I noticed: this restriction (not allowing duplicate environments), should not only be applied when updating a consumer, but also when registering a consumer. we should add this restriction in ConsumerResource and also enhance/add this test case here either in this same test or a separate test

Reply:
Done. Added validations on creation as well.